--- a/apple-ib-drv/apple-touchbar.c
+++ b/apple-ib-drv/apple-touchbar.c
@@ -1405,8 +1405,8 @@ static int appletb_reset_resume(struct hid_device *hdev)
 	if (iface_info)
 		iface_info->suspended = false;
 
-	if ((tb_dev->mode_iface.hdev && !tb_dev->mode_iface.suspended) &&
-	    (tb_dev->disp_iface.hdev && !tb_dev->disp_iface.suspended)) {
+	if ((!tb_dev->mode_iface.hdev || !tb_dev->mode_iface.suspended) &&
+	    (!tb_dev->disp_iface.hdev || !tb_dev->disp_iface.suspended)) {
 		/*
 		 * Restore touch bar state. Note that autopm state is not
 		 * preserved, so need explicitly restore that here.
@@ -1471,6 +1471,7 @@ static struct hid_driver appletb_hid_driver = {
 	.input_configured = appletb_input_configured,
 #ifdef CONFIG_PM
 	.suspend = appletb_suspend,
+	.resume = appletb_reset_resume,
 	.reset_resume = appletb_reset_resume,
 #endif
 };
