diff --git a/src/dkms.conf b/src/dkms.conf
index 78eaf8b..9d93675 100644
--- a/src/dkms.conf
+++ b/src/dkms.conf
@@ -3,6 +3,3 @@
-CLEAN="make clean"
-MAKE[0]="make"
 BUILT_MODULE_NAME[0]="yt6801"
 DEST_MODULE_LOCATION[0]="/kernel/drivers/net/ethernet/motorcomm"
 AUTOINSTALL="yes"
-REMAKE_INITRD="yes"
diff --git a/src/fuxi-gmac-net.c b/src/fuxi-gmac-net.c
index 3e643f9..fea4c97 100644
--- a/src/fuxi-gmac-net.c
+++ b/src/fuxi-gmac-net.c
@@ -771,7 +771,9 @@ static void fxgmac_tx_hang_timer_handler(struct timer_list *t)
 static void fxgmac_tx_hang_timer_handler(unsigned long data)
 #endif
 {
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,14,0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,16,0))
+    struct fxgmac_channel *channel = timer_container_of(channel, t, expansion.tx_hang_timer);
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(4,14,0))
     struct fxgmac_channel *channel = from_timer(channel, t, expansion.tx_hang_timer);
 #else
     struct fxgmac_channel *channel = (struct fxgmac_channel *)data;
diff --git a/src/fuxi-gmac.h b/src/fuxi-gmac.h
index 539b507..2435f0c 100644
--- a/src/fuxi-gmac.h
+++ b/src/fuxi-gmac.h
@@ -50,6 +50,8 @@
 #define FXGMAC_NS_IFA_LOCAL_LINK     1
 #define FXGMAC_NS_IFA_GLOBAL_UNICAST 2
 
+#define FXGMAC_INT_MODERATION_ENABLED 1
+
 #define FXGMAX_ASPM_WAR_EN
 /* Descriptor related parameters */
 #if FXGMAC_TX_HANG_TIMER_ENABLED
