--- a/PKGBUILD
+++ b/PKGBUILD
@@ -36,7 +36,6 @@ depends=(
   'libcap'
   'libdrm'
   'libevdev'
-  'libmfx'
   'libnotify'
   'libpipewire'
   'libpulse'
@@ -53,6 +52,10 @@ depends=(
   'which'
 )
 
+depends_x86_64=(
+  'libmfx'
+)
+
 makedepends=(
   'appstream'
   'appstream-glib'
@@ -69,6 +72,10 @@ makedepends=(
   'shaderc'
 )
 
+makedepends_aarch64=(
+  'vulkan-headers'
+)
+
 checkdepends=(
   'gcovr'
 )
@@ -111,7 +118,12 @@ if [ -n "${optdepends+x}" ]; then
 
 prepare() {
     cd "$pkgname"
-    git submodule update --recursive --init
+    if [[ $CARCH == aarch64 ]]; then
+      git submodule update --init
+      git -C third-party/moonlight-common-c submodule update --init enet
+    else
+      git submodule update --recursive --init
+    fi
 
     # Backport https://github.com/LizardByte/Sunshine/commit/060b6b07d376e9ef4c4ce4a5c6d648d08fe0eab9
     patch -Np1 -i "${srcdir}/2026.516.143833-060b6b07-remove-setuptools-from-glad-dependencies.patch"
@@ -145,6 +157,10 @@ build() {
       -D SUNSHINE_PUBLISHER_ISSUE_URL='https://app.lizardbyte.dev/support'
     )
 
+    if [[ $CARCH == aarch64 ]]; then
+      _cmake_options+=(-D SUNSHINE_SYSTEM_VULKAN_HEADERS=ON)
+    fi
+
     if [[ "${_use_cuda::1}" != "t" ]]; then
       _cmake_options+=(-DSUNSHINE_ENABLE_CUDA=OFF -DCUDA_FAIL_ON_MISSING=OFF)
     else
