diff --git a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
--- a/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
+++ b/drivers/gpu/drm/amd/display/amdgpu_dm/amdgpu_dm.c
@@ -10135,20 +10135,19 @@ static void update_freesync_state_on_stream(
 	 */
 	if (new_stream->signal == SIGNAL_TYPE_HDMI_TYPE_A ||
 	    new_stream->signal == SIGNAL_TYPE_HDMI_FRL) {
-		struct dc_info_packet hf_vsif = {0};
+		struct dc_info_packet vsp_infopacket = {0};
 		bool sink_allm = aconn && aconn->base.display_info.hdmi.allm;
 		bool allm = sink_allm &&
 			(vrr_params.state == VRR_STATE_ACTIVE_VARIABLE ||
 			 vrr_params.state == VRR_STATE_ACTIVE_FIXED);
 		bool allm_changed;
 
-		if (allm)
-			mod_build_hf_vsif_infopacket(new_stream, &hf_vsif, allm, allm);
+		mod_build_hf_vsif_infopacket(new_stream, &vsp_infopacket, allm, allm);
 
-		allm_changed = memcmp(&new_stream->hfvsif_infopacket, &hf_vsif,
-				      sizeof(hf_vsif)) != 0;
+		allm_changed = memcmp(&new_stream->vsp_infopacket, &vsp_infopacket,
+				      sizeof(vsp_infopacket)) != 0;
 		new_crtc_state->freesync_vrr_info_changed |= allm_changed;
-		new_stream->hfvsif_infopacket = hf_vsif;
+		new_stream->vsp_infopacket = vsp_infopacket;
 
 		if (allm_changed)
 			drm_dbg_driver(adev_to_drm(adev),
@@ -10727,8 +10726,8 @@ static void amdgpu_dm_commit_planes(struct drm_atomic_commit *state,
 			if (acrtc_state->freesync_vrr_info_changed) {
 				bundle->stream_update.vrr_infopacket =
 					&acrtc_state->stream->vrr_infopacket;
-				bundle->stream_update.hfvsif_infopacket =
-					&acrtc_state->stream->hfvsif_infopacket;
+				bundle->stream_update.vsp_infopacket =
+					&acrtc_state->stream->vsp_infopacket;
 			}
 		}
 	}
