diff --git a/PKGBUILD b/PKGBUILD
index d560ca4..f149827 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -13,14 +13,13 @@ license=('GPL-2.0-only')
 arch=('any')
 
-makedepends=('git')
 optdepends=('facetimehd-data: Sensor calibration data')
 
 provides=('bcwc-pcie' 'bcwc-pcie-dkms')
 conflicts=('bcwc-pcie' 'bcwc-pcie-dkms')
 
-_pkgsrc="$_pkgname"
-source=("$_pkgsrc"::"git+https://github.com/patjak/facetimehd.git#tag=$pkgver")
-sha256sums=('e25dbbf89f2145c8a61ee53a594271ef5bb3a6b8b006e70c637996240fffc219')
+_pkgsrc="facetimehd-$pkgver"
+source=("$_pkgsrc.tar.gz::https://github.com/patjak/facetimehd/archive/refs/tags/$pkgver.tar.gz")
+sha256sums=('3bf7cdfa62c57258a1f6188caf10e767d9315574deefaf246d4725a17547191a')
 
 prepare() {
   # deprecated
