diff --git a/PKGBUILD b/PKGBUILD
index f149827..1ca9c20 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -13,6 +13,10 @@
 license=('GPL-2.0-only')
 arch=('any')
 
+depends=(
+  'facetimehd-firmware'
+  'dkms'
+)
 optdepends=('facetimehd-data: Sensor calibration data')
 
 provides=('bcwc-pcie' 'bcwc-pcie-dkms')
@@ -28,11 +32,6 @@
 }
 
 package() {
-  depends=(
-    'facetimehd-firmware'
-    'dkms'
-  )
-
   cd "$_pkgsrc"
   for FILE in dkms.conf Makefile *.[ch]; do
     install -Dm644 "$FILE" "$pkgdir/usr/src/facetimehd-$pkgver/$FILE"
