diff --git a/tests/meson.build b/tests/meson.build
index 1f2982d..8dcde82 100644
--- a/tests/meson.build
+++ b/tests/meson.build
@@ -86,7 +86,7 @@ test_emulation = configuration_data()
 test_emulation.set_quoted('FPI_EMULATION_ENV_VAR',
   fp_device_emulation_env_var)
 test_emulation.set_quoted('FPI_EMULATION_HELPER_BUILDDIR',
-  meson.project_build_root() / 'tests')
+  'tests')
 test_emulation.set_quoted('FPI_EMULATION_HELPER_INSTALLDIR',
   libexecdir / 'installed-tests' / versioned_libname)
 test_emulation.set_quoted('FPI_EMULATION_HELPER_MODULE',
